Failed to bind TCP server to port 0.0.0.0:8443 due to io.netty.channel.unix.Errors$NativeIoException: bind(..) failed: Address already in use

My issue is a followup of https://github.com/zeebe-io/zeebe/issues/5410. The answer to that issue helped me set up advertised addresses and hosts for command, internal and monitoring APIs (I'm mostly interested in the first two though).

After having set that up accordingly, the broker crashes because 0.0.0.0:8443 is in use by another service. 8443 is a proxy port, not a port that the broker has to be listening to. This is very strange behavior. Please look at my log below. It shows the application.yml settings and the errors I'm getting. Thank you for the help.
```
$ kubectl logs pod/zeebe-cluster-0 -n 103462-zeebe-2-dev
Using startup_fromconfmap.sh as startup script.
This is updated by a rolling restart.
+ echo 'Using startup_fromconfmap.sh as startup script.'
+ echo 'This is updated by a rolling restart.'
+ export ZEEBE_BROKER_CLUSTER_NODEID=0
+ ZEEBE_BROKER_CLUSTER_NODEID=0
+ ZEEBE_NODE_ID=0
++ hostname -f
+ export ZEEBE_BROKER_NETWORK_HOST=zeebe-cluster-0.zeebe-cluster.103462-zeebe-2-dev.svc.cluster.local
+ ZEEBE_BROKER_NETWORK_HOST=zeebe-cluster-0.zeebe-cluster.103462-zeebe-2-dev.svc.cluster.local
+ export ZEEBE_BROKER_NETWORK_COMMANDAPI_ADVERTISEDHOST=zeebe-broker-0-command.apps.mt-d1.pool1.proxydomain.net
+ ZEEBE_BROKER_NETWORK_COMMANDAPI_ADVERTISEDHOST=zeebe-broker-0-command.apps.mt-d1.pool1.proxydomain.net
+ export ZEEBE_BROKER_NETWORK_COMMANDAPI_ADVERTISEDPORT=8443
+ ZEEBE_BROKER_NETWORK_COMMANDAPI_ADVERTISEDPORT=8443
+ export ZEEBE_BROKER_NETWORK_INTERNALAPI_ADVERTISEDHOST=zeebe-broker-0-internal.apps.mt-d1.pool1.proxydomain.net
+ ZEEBE_BROKER_NETWORK_INTERNALAPI_ADVERTISEDHOST=zeebe-broker-0-internal.apps.mt-d1.pool1.proxydomain.net
+ export ZEEBE_BROKER_NETWORK_INTERNALAPI_ADVERTISEDPORT=8443
+ ZEEBE_BROKER_NETWORK_INTERNALAPI_ADVERTISEDPORT=8443
+ echo 'ZEEBE_STANDALONE_GATEWAY is false'
ZEEBE_STANDALONE_GATEWAY is false
+ '[' false = true ']'
+ ZEEBE_BROKER_CLUSTER_INITIALCONTACTPOINTS=zeebe-cluster-0.zeebe-cluster.103462-zeebe-2-dev.svc.cluster.local:26502
+ (( i=0 ))
+ (( i<6/2 ))
+ '[' 0 '!=' 0 ']'
+ (( i++  ))
+ (( i<6/2 ))
+ '[' 0 '!=' 1 ']'
++ hostname -d
+ ZEEBE_BROKER_CLUSTER_INITIALCONTACTPOINTS=zeebe-cluster-0.zeebe-cluster.103462-zeebe-2-dev.svc.cluster.local:26502,zeebe-cluster-1.zeebe-cluster.103462-zeebe-2-dev.svc.cluster.local:26502
+ (( i++  ))
+ (( i<6/2 ))
+ '[' 0 '!=' 2 ']'
++ hostname -d
+ ZEEBE_BROKER_CLUSTER_INITIALCONTACTPOINTS=zeebe-cluster-0.zeebe-cluster.103462-zeebe-2-dev.svc.cluster.local:26502,zeebe-cluster-1.zeebe-cluster.103462-zeebe-2-dev.svc.cluster.local:26502,zeebe-cluster-2.zeebe-cluster.103462-zeebe-2-dev.svc.cluster.local:26502
+ (( i++  ))
+ (( i<6/2 ))
+ REMOTE_BROKER_0=zeebe-broker-0-internal.apps.mt-d1.pool2.proxydomain.net:8443
+ REMOTE_BROKER_1=zeebe-broker-1-internal.apps.mt-d1.pool2.proxydomain.net:8443
+ REMOTE_BROKER_2=zeebe-broker-2-internal.apps.mt-d1.pool2.proxydomain.net:8443
+ ZEEBE_BROKER_CLUSTER_INITIALCONTACTPOINTS=zeebe-cluster-0.zeebe-cluster.103462-zeebe-2-dev.svc.cluster.local:26502,zeebe-cluster-1.zeebe-cluster.103462-zeebe-2-dev.svc.cluster.local:26502,zeebe-cluster-2.zeebe-cluster.103462-zeebe-2-dev.svc.cluster.local:26502,zeebe-broker-0-internal.apps.mt-d1.pool2.proxydomain.net:8443,zeebe-broker-1-internal.apps.mt-d1.pool2.proxydomain.net:8443,zeebe-broker-2-internal.apps.mt-d1.pool2.proxydomain.net:8443
+ export ZEEBE_BROKER_CLUSTER_INITIALCONTACTPOINTS=zeebe-cluster-0.zeebe-cluster.103462-zeebe-2-dev.svc.cluster.local:26502,zeebe-cluster-1.zeebe-cluster.103462-zeebe-2-dev.svc.cluster.local:26502,zeebe-cluster-2.zeebe-cluster.103462-zeebe-2-dev.svc.cluster.local:26502,zeebe-broker-0-internal.apps.mt-d1.pool2.proxydomain.net:8443,zeebe-broker-1-internal.apps.mt-d1.pool2.proxydomain.net:8443,zeebe-broker-2-internal.apps.mt-d1.pool2.proxydomain.net:8443
+ ZEEBE_BROKER_CLUSTER_INITIALCONTACTPOINTS=zeebe-cluster-0.zeebe-cluster.103462-zeebe-2-dev.svc.cluster.local:26502,zeebe-cluster-1.zeebe-cluster.103462-zeebe-2-dev.svc.cluster.local:26502,zeebe-cluster-2.zeebe-cluster.103462-zeebe-2-dev.svc.cluster.local:26502,zeebe-broker-0-internal.apps.mt-d1.pool2.proxydomain.net:8443,zeebe-broker-1-internal.apps.mt-d1.pool2.proxydomain.net:8443,zeebe-broker-2-internal.apps.mt-d1.pool2.proxydomain.net:8443
+ exec /usr/local/zeebe/bin/broker
  ______  ______   ______   ____    ______     ____    _____     ____    _  __  ______   _____
 |___  / |  ____| |  ____| |  _ \  |  ____|   |  _ \  |  __ \   / __ \  | |/ / |  ____| |  __ \
    / /  | |__    | |__    | |_) | | |__      | |_) | | |__) | | |  | | | ' /  | |__    | |__) |
   / /   |  __|   |  __|   |  _ <  |  __|     |  _ <  |  _  /  | |  | | |  <   |  __|   |  _  /
  / /__  | |____  | |____  | |_) | | |____    | |_) | | | \ \  | |__| | | . \  | |____  | | \ \
 /_____| |______| |______| |____/  |______|   |____/  |_|  \_\  \____/  |_|\_\ |______| |_|  \_\

2020-09-24 13:53:21.495 [] [main] INFO  io.zeebe.broker.StandaloneBroker - Starting StandaloneBroker v0.23.2 on zeebe-cluster-0 with PID 1 (/usr/local/zeebe/lib/zeebe-distribution-0.23.2.jar started by ? in /usr/local/zeebe)
2020-09-24 13:53:21.499 [] [main] INFO  io.zeebe.broker.StandaloneBroker - No active profile set, falling back to default profiles: default
2020-09-24 13:53:22.413 [] [main] INFO  io.zeebe.broker.StandaloneBroker - Started StandaloneBroker in 1.341 seconds (JVM running for 2.432)
2020-09-24 13:53:22.480 [] [main] INFO  io.zeebe.broker.system - Version: 0.23.2
2020-09-24 13:53:22.581 [] [main] INFO  io.zeebe.broker.system - Starting broker 0 with configuration {
  "network" : {
    "host" : "zeebe-cluster-0.zeebe-cluster.103462-zeebe-2-dev.svc.cluster.local",
    "portOffset" : 0,
    "maxMessageSize" : "4MB",
    "advertisedHost" : "zeebe-cluster-0.zeebe-cluster.103462-zeebe-2-dev.svc.cluster.local",
    "commandApi" : {
      "host" : "zeebe-cluster-0.zeebe-cluster.103462-zeebe-2-dev.svc.cluster.local",
      "port" : 26501,
      "advertisedHost" : "zeebe-broker-0-command.apps.mt-d1.pool1.proxydomain.net",
      "advertisedPort" : 8443,
      "address" : "zeebe-cluster-0.zeebe-cluster.103462-zeebe-2-dev.svc.cluster.local:26501",
      "advertisedAddress" : "zeebe-broker-0-command.apps.mt-d1.pool1.proxydomain.net:8443"
    },
    "internalApi" : {
      "host" : "zeebe-cluster-0.zeebe-cluster.103462-zeebe-2-dev.svc.cluster.local",
      "port" : 26502,
      "advertisedHost" : "zeebe-broker-0-internal.apps.mt-d1.pool1.proxydomain.net",
      "advertisedPort" : 8443,
      "address" : "zeebe-cluster-0.zeebe-cluster.103462-zeebe-2-dev.svc.cluster.local:26502",
      "advertisedAddress" : "zeebe-broker-0-internal.apps.mt-d1.pool1.proxydomain.net:8443"
    },
    "monitoringApi" : {
      "host" : "zeebe-cluster-0.zeebe-cluster.103462-zeebe-2-dev.svc.cluster.local",
      "port" : 9600,
      "advertisedHost" : "zeebe-cluster-0.zeebe-cluster.103462-zeebe-2-dev.svc.cluster.local",
      "advertisedPort" : 9600,
      "address" : "zeebe-cluster-0.zeebe-cluster.103462-zeebe-2-dev.svc.cluster.local:9600",
      "advertisedAddress" : "zeebe-cluster-0.zeebe-cluster.103462-zeebe-2-dev.svc.cluster.local:9600"
    },
    "maxMessageSizeInBytes" : 4194304
  },
  "cluster" : {
    "initialContactPoints" : [ "zeebe-cluster-0.zeebe-cluster.103462-zeebe-2-dev.svc.cluster.local:26502", "zeebe-cluster-1.zeebe-cluster.103462-zeebe-2-dev.svc.cluster.local:26502", "zeebe-cluster-2.zeebe-cluster.103462-zeebe-2-dev.svc.cluster.local:26502", "zeebe-broker-0-internal.apps.mt-d1.pool2.proxydomain.net:8443", "zeebe-broker-1-internal.apps.mt-d1.pool2.proxydomain.net:8443", "zeebe-broker-2-internal.apps.mt-d1.pool2.proxydomain.net:8443" ],
    "partitionIds" : [ 1, 2 ],
    "nodeId" : 0,
    "partitionsCount" : 2,
    "replicationFactor" : 3,
    "clusterSize" : 6,
    "clusterName" : "zeebe-cluster",
    "gossipFailureTimeout" : 10000,
    "gossipInterval" : 250,
    "gossipProbeInterval" : 1000
  },
  "threads" : {
    "cpuThreadCount" : 2,
    "ioThreadCount" : 2
  },
  "data" : {
    "directories" : [ "/usr/local/zeebe/data" ],
    "logSegmentSize" : "512MB",
    "snapshotPeriod" : "PT15M",
    "logIndexDensity" : 100,
    "logSegmentSizeInBytes" : 536870912,
    "atomixStorageLevel" : "DISK"
  },
  "exporters" : {
    "elasticsearch" : {
      "jarPath" : null,
      "className" : "io.zeebe.exporter.ElasticsearchExporter",
      "args" : {
        "url" : "http://zeebe-elasticsearch-service:9200",
        "bulk" : {
          "delay" : 5,
          "size" : 1000
        },
        "index" : {
          "prefix" : "zeebe-record",
          "createTemplate" : true,
          "command" : false,
          "event" : true,
          "rejection" : false,
          "deployment" : true,
          "error" : true,
          "incident" : true,
          "job" : true,
          "jobBatch" : false,
          "message" : false,
          "messageSubscription" : false,
          "variable" : true,
          "variableDocument" : true,
          "workflowInstance" : true,
          "workflowInstanceCreation" : false,
          "workflowInstanceSubscription" : false,
          "ignoreVariablesAbove" : 32677
        }
      },
      "external" : false
    }
  },
  "gateway" : {
    "network" : {
      "host" : "0.0.0.0",
      "port" : 26500,
      "minKeepAliveInterval" : "PT30S"
    },
    "cluster" : {
      "contactPoint" : "zeebe-cluster-0.zeebe-cluster.103462-zeebe-2-dev.svc.cluster.local:26502",
      "requestTimeout" : "PT15S",
      "clusterName" : "zeebe-cluster",
      "memberId" : "gateway",
      "host" : "0.0.0.0",
      "port" : 26502
    },
    "threads" : {
      "managementThreads" : 1
    },
    "monitoring" : {
      "enabled" : false,
      "host" : "zeebe-cluster-0.zeebe-cluster.103462-zeebe-2-dev.svc.cluster.local",
      "port" : 9600
    },
    "security" : {
      "enabled" : false,
      "certificateChainPath" : null,
      "privateKeyPath" : null
    },
    "enable" : true
  },
  "backpressure" : {
    "enabled" : true,
    "algorithm" : "VEGAS"
  },
  "stepTimeout" : "PT5M"
}
2020-09-24 13:53:22.599 [] [main] INFO  io.zeebe.broker.system - Bootstrap Broker-0 [1/11]: actor scheduler
2020-09-24 13:53:22.601 [] [main] INFO  io.zeebe.broker.system - Bootstrap Broker-0 [2/11]: membership and replication protocol
2020-09-24 13:53:24.847 [] [main] INFO  io.zeebe.broker.system - Bootstrap Broker-0 [3/11]: command api transport
2020-09-24 13:53:25.138 [] [main] INFO  io.zeebe.broker.system - Bootstrap Broker-0 [4/11]: command api handler
2020-09-24 13:53:25.170 [] [main] INFO  io.zeebe.broker.system - Bootstrap Broker-0 [5/11]: subscription api
2020-09-24 13:53:25.189 [] [main] INFO  io.zeebe.broker.system - Bootstrap Broker-0 [6/11]: embedded gateway
2020-09-24 13:53:25.194 [] [main] INFO  io.zeebe.gateway - Version: 0.23.2
2020-09-24 13:53:25.195 [] [main] INFO  io.zeebe.gateway - Starting gateway with configuration {
  "network" : {
    "host" : "0.0.0.0",
    "port" : 26500,
    "minKeepAliveInterval" : "PT30S"
  },
  "cluster" : {
    "contactPoint" : "zeebe-cluster-0.zeebe-cluster.103462-zeebe-2-dev.svc.cluster.local:26502",
    "requestTimeout" : "PT15S",
    "clusterName" : "zeebe-cluster",
    "memberId" : "gateway",
    "host" : "0.0.0.0",
    "port" : 26502
  },
  "threads" : {
    "managementThreads" : 1
  },
  "monitoring" : {
    "enabled" : false,
    "host" : "zeebe-cluster-0.zeebe-cluster.103462-zeebe-2-dev.svc.cluster.local",
    "port" : 9600
  },
  "security" : {
    "enabled" : false,
    "certificateChainPath" : null,
    "privateKeyPath" : null
  },
  "enable" : true
}
2020-09-24 13:53:25.298 [] [main] INFO  io.zeebe.broker.system - Bootstrap Broker-0 [7/11]: cluster services
2020-09-24 13:53:25.302 [] [netty-messaging-event-epoll-server-0] WARN  io.atomix.cluster.messaging.impl.NettyMessagingService - Failed to bind TCP server to port 0.0.0.0:8443 due to io.netty.channel.unix.Errors$NativeIoException: bind(..) failed: Address already in use
2020-09-24 13:53:25.303 [] [main] INFO  io.zeebe.broker.system - Bootstrap Broker-0 [7/11]: cluster services failed with unexpected exception.
java.util.concurrent.CompletionException: io.netty.channel.unix.Errors$NativeIoException: bind(..) failed: Address already in use
        at java.util.concurrent.CompletableFuture.encodeThrowable(Unknown Source) ~[?:?]
        at java.util.concurrent.CompletableFuture.completeThrowable(Unknown Source) ~[?:?]
        at java.util.concurrent.CompletableFuture$UniRun.tryFire(Unknown Source) ~[?:?]
        at java.util.concurrent.CompletableFuture.postComplete(Unknown Source) ~[?:?]
        at java.util.concurrent.CompletableFuture.completeExceptionally(Unknown Source) ~[?:?]
        at io.atomix.cluster.messaging.impl.NettyMessagingService.lambda$bind$32(NettyMessagingService.java:699) ~[atomix-cluster-0.23.2.jar:0.23.2]
        at io.netty.util.concurrent.DefaultPromise.notifyListener0(DefaultPromise.java:577) ~[netty-common-4.1.50.Final.jar:4.1.50.Final]
        at io.netty.util.concurrent.DefaultPromise.notifyListenersNow(DefaultPromise.java:551) ~[netty-common-4.1.50.Final.jar:4.1.50.Final]
        at io.netty.util.concurrent.DefaultPromise.notifyListeners(DefaultPromise.java:490) ~[netty-common-4.1.50.Final.jar:4.1.50.Final]
        at io.netty.util.concurrent.DefaultPromise.setValue0(DefaultPromise.java:615) ~[netty-common-4.1.50.Final.jar:4.1.50.Final]
        at io.netty.util.concurrent.DefaultPromise.setFailure0(DefaultPromise.java:608) ~[netty-common-4.1.50.Final.jar:4.1.50.Final]
        at io.netty.util.concurrent.DefaultPromise.tryFailure(DefaultPromise.java:117) ~[netty-common-4.1.50.Final.jar:4.1.50.Final]
        at io.netty.channel.AbstractChannel$AbstractUnsafe.safeSetFailure(AbstractChannel.java:993) ~[netty-transport-4.1.50.Final.jar:4.1.50.Final]
        at io.netty.channel.AbstractChannel$AbstractUnsafe.bind(AbstractChannel.java:552) ~[netty-transport-4.1.50.Final.jar:4.1.50.Final]
        at io.netty.channel.DefaultChannelPipeline$HeadContext.bind(DefaultChannelPipeline.java:1334) ~[netty-transport-4.1.50.Final.jar:4.1.50.Final]
        at io.netty.channel.AbstractChannelHandlerContext.invokeBind(AbstractChannelHandlerContext.java:506) ~[netty-transport-4.1.50.Final.jar:4.1.50.Final]
        at io.netty.channel.AbstractChannelHandlerContext.bind(AbstractChannelHandlerContext.java:491) ~[netty-transport-4.1.50.Final.jar:4.1.50.Final]
        at io.netty.channel.DefaultChannelPipeline.bind(DefaultChannelPipeline.java:973) ~[netty-transport-4.1.50.Final.jar:4.1.50.Final]
        at io.netty.channel.AbstractChannel.bind(AbstractChannel.java:248) ~[netty-transport-4.1.50.Final.jar:4.1.50.Final]
        at io.netty.bootstrap.AbstractBootstrap$2.run(AbstractBootstrap.java:356) ~[netty-transport-4.1.50.Final.jar:4.1.50.Final]
        at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:164) ~[netty-common-4.1.50.Final.jar:4.1.50.Final]
        at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:472) ~[netty-common-4.1.50.Final.jar:4.1.50.Final]
        at io.netty.channel.epoll.EpollEventLoop.run(EpollEventLoop.java:387) ~[netty-transport-native-epoll-4.1.50.Final-linux-x86_64.jar:4.1.50.Final]
        at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989) ~[netty-common-4.1.50.Final.jar:4.1.50.Final]
        at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74) ~[netty-common-4.1.50.Final.jar:4.1.50.Final]
        at java.lang.Thread.run(Unknown Source) ~[?:?]
Caused by: io.netty.channel.unix.Errors$NativeIoException: bind(..) failed: Address already in use
2020-09-24 13:53:25.307 [] [main] INFO  io.zeebe.broker.system - Closing Broker-0 [1/6]: embedded gateway
2020-09-24 13:53:25.313 [] [main] INFO  io.zeebe.broker.system - Closing Broker-0 [2/6]: subscription api
2020-09-24 13:53:25.314 [] [main] INFO  io.zeebe.broker.system - Closing Broker-0 [3/6]: command api handler
2020-09-24 13:53:25.314 [] [main] INFO  io.zeebe.broker.system - Closing Broker-0 [4/6]: command api transport
2020-09-24 13:53:27.329 [] [main] INFO  io.zeebe.broker.system - Closing Broker-0 [5/6]: membership and replication protocol
2020-09-24 13:53:27.331 [] [main] WARN  io.atomix.utils.event.ListenerRegistry - Listener io.atomix.primitive.partition.impl.DefaultPartitionService$$Lambda$529/0x000000080056c440@3de8f85c not registered
2020-09-24 13:53:27.334 [] [main] WARN  io.atomix.utils.event.ListenerRegistry - Listener io.atomix.primitive.partition.impl.DefaultPartitionGroupMembershipService$$Lambda$530/0x000000080056b840@359e27d2 not registered
2020-09-24 13:53:27.338 [] [main] INFO  io.zeebe.broker.system - Closing Broker-0 [6/6]: actor scheduler
2020-09-24 13:53:27.340 [] [main] INFO  io.zeebe.broker.system - Closing Broker-0 succeeded. Closed 6 steps in 2033 ms.
2020-09-24 13:53:27.340 [] [main] ERROR io.zeebe.broker.system - Failed to start broker 0!
java.util.concurrent.CompletionException: io.netty.channel.unix.Errors$NativeIoException: bind(..) failed: Address already in use
        at java.util.concurrent.CompletableFuture.encodeThrowable(Unknown Source) ~[?:?]
        at java.util.concurrent.CompletableFuture.completeThrowable(Unknown Source) ~[?:?]
        at java.util.concurrent.CompletableFuture$UniRun.tryFire(Unknown Source) ~[?:?]
        at java.util.concurrent.CompletableFuture.postComplete(Unknown Source) ~[?:?]
        at java.util.concurrent.CompletableFuture.completeExceptionally(Unknown Source) ~[?:?]
        at io.atomix.cluster.messaging.impl.NettyMessagingService.lambda$bind$32(NettyMessagingService.java:699) ~[atomix-cluster-0.23.2.jar:0.23.2]
        at io.netty.util.concurrent.DefaultPromise.notifyListener0(DefaultPromise.java:577) ~[netty-common-4.1.50.Final.jar:4.1.50.Final]
        at io.netty.util.concurrent.DefaultPromise.notifyListenersNow(DefaultPromise.java:551) ~[netty-common-4.1.50.Final.jar:4.1.50.Final]
        at io.netty.util.concurrent.DefaultPromise.notifyListeners(DefaultPromise.java:490) ~[netty-common-4.1.50.Final.jar:4.1.50.Final]
        at io.netty.util.concurrent.DefaultPromise.setValue0(DefaultPromise.java:615) ~[netty-common-4.1.50.Final.jar:4.1.50.Final]
        at io.netty.util.concurrent.DefaultPromise.setFailure0(DefaultPromise.java:608) ~[netty-common-4.1.50.Final.jar:4.1.50.Final]
        at io.netty.util.concurrent.DefaultPromise.tryFailure(DefaultPromise.java:117) ~[netty-common-4.1.50.Final.jar:4.1.50.Final]
        at io.netty.channel.AbstractChannel$AbstractUnsafe.safeSetFailure(AbstractChannel.java:993) ~[netty-transport-4.1.50.Final.jar:4.1.50.Final]
        at io.netty.channel.AbstractChannel$AbstractUnsafe.bind(AbstractChannel.java:552) ~[netty-transport-4.1.50.Final.jar:4.1.50.Final]
        at io.netty.channel.DefaultChannelPipeline$HeadContext.bind(DefaultChannelPipeline.java:1334) ~[netty-transport-4.1.50.Final.jar:4.1.50.Final]
        at io.netty.channel.AbstractChannelHandlerContext.invokeBind(AbstractChannelHandlerContext.java:506) ~[netty-transport-4.1.50.Final.jar:4.1.50.Final]
        at io.netty.channel.AbstractChannelHandlerContext.bind(AbstractChannelHandlerContext.java:491) ~[netty-transport-4.1.50.Final.jar:4.1.50.Final]
        at io.netty.channel.DefaultChannelPipeline.bind(DefaultChannelPipeline.java:973) ~[netty-transport-4.1.50.Final.jar:4.1.50.Final]
        at io.netty.channel.AbstractChannel.bind(AbstractChannel.java:248) ~[netty-transport-4.1.50.Final.jar:4.1.50.Final]
        at io.netty.bootstrap.AbstractBootstrap$2.run(AbstractBootstrap.java:356) ~[netty-transport-4.1.50.Final.jar:4.1.50.Final]
        at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:164) ~[netty-common-4.1.50.Final.jar:4.1.50.Final]
        at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:472) ~[netty-common-4.1.50.Final.jar:4.1.50.Final]
        at io.netty.channel.epoll.EpollEventLoop.run(EpollEventLoop.java:387) ~[netty-transport-native-epoll-4.1.50.Final-linux-x86_64.jar:4.1.50.Final]
        at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989) ~[netty-common-4.1.50.Final.jar:4.1.50.Final]
        at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74) ~[netty-common-4.1.50.Final.jar:4.1.50.Final]
        at java.lang.Thread.run(Unknown Source) ~[?:?]
Caused by: io.netty.channel.unix.Errors$NativeIoException: bind(..) failed: Address already in use
2020-09-24 13:53:27.343 [] [main] INFO  org.springframework.boot.autoconfigure.logging.ConditionEvaluationReportLoggingListener -
Error starting ApplicationContext. To display the conditions report re-run your application with 'debug' enabled.
2020-09-24 13:53:27.345 [] [main] ERROR org.springframework.boot.SpringApplication - Application run failed
java.lang.IllegalStateException: Failed to execute CommandLineRunner
        at org.springframework.boot.SpringApplication.callRunner(SpringApplication.java:787) [spring-boot-2.2.5.RELEASE.jar:2.2.5.RELEASE]
        at org.springframework.boot.SpringApplication.callRunners(SpringApplication.java:768) [spring-boot-2.2.5.RELEASE.jar:2.2.5.RELEASE]
        at org.springframework.boot.SpringApplication.run(SpringApplication.java:322) [spring-boot-2.2.5.RELEASE.jar:2.2.5.RELEASE]
        at org.springframework.boot.SpringApplication.run(SpringApplication.java:1226) [spring-boot-2.2.5.RELEASE.jar:2.2.5.RELEASE]
        at org.springframework.boot.SpringApplication.run(SpringApplication.java:1215) [spring-boot-2.2.5.RELEASE.jar:2.2.5.RELEASE]
        at io.zeebe.broker.StandaloneBroker.main(StandaloneBroker.java:46) [zeebe-distribution-0.23.2.jar:0.23.2]
Caused by: io.zeebe.util.exception.UncheckedExecutionException: Failed to start broker
        at io.zeebe.broker.Broker.internalStart(Broker.java:142) ~[zeebe-broker-0.23.2.jar:0.23.2]
        at io.zeebe.util.LogUtil.doWithMDC(LogUtil.java:21) ~[zeebe-util-0.23.2.jar:0.23.2]
        at io.zeebe.broker.Broker.start(Broker.java:115) ~[zeebe-broker-0.23.2.jar:0.23.2]
        at io.zeebe.broker.StandaloneBroker.run(StandaloneBroker.java:59) ~[zeebe-distribution-0.23.2.jar:0.23.2]
        at org.springframework.boot.SpringApplication.callRunner(SpringApplication.java:784) ~[spring-boot-2.2.5.RELEASE.jar:2.2.5.RELEASE]
        ... 5 more
Caused by: java.util.concurrent.CompletionException: io.netty.channel.unix.Errors$NativeIoException: bind(..) failed: Address already in use
        at java.util.concurrent.CompletableFuture.encodeThrowable(Unknown Source) ~[?:?]
        at java.util.concurrent.CompletableFuture.completeThrowable(Unknown Source) ~[?:?]
        at java.util.concurrent.CompletableFuture$UniRun.tryFire(Unknown Source) ~[?:?]
        at java.util.concurrent.CompletableFuture.postComplete(Unknown Source) ~[?:?]
        at java.util.concurrent.CompletableFuture.completeExceptionally(Unknown Source) ~[?:?]
        at io.atomix.cluster.messaging.impl.NettyMessagingService.lambda$bind$32(NettyMessagingService.java:699) ~[atomix-cluster-0.23.2.jar:0.23.2]
        at io.netty.util.concurrent.DefaultPromise.notifyListener0(DefaultPromise.java:577) ~[netty-common-4.1.50.Final.jar:4.1.50.Final]
        at io.netty.util.concurrent.DefaultPromise.notifyListenersNow(DefaultPromise.java:551) ~[netty-common-4.1.50.Final.jar:4.1.50.Final]
        at io.netty.util.concurrent.DefaultPromise.notifyListeners(DefaultPromise.java:490) ~[netty-common-4.1.50.Final.jar:4.1.50.Final]
        at io.netty.util.concurrent.DefaultPromise.setValue0(DefaultPromise.java:615) ~[netty-common-4.1.50.Final.jar:4.1.50.Final]
        at io.netty.util.concurrent.DefaultPromise.setFailure0(DefaultPromise.java:608) ~[netty-common-4.1.50.Final.jar:4.1.50.Final]
        at io.netty.util.concurrent.DefaultPromise.tryFailure(DefaultPromise.java:117) ~[netty-common-4.1.50.Final.jar:4.1.50.Final]
        at io.netty.channel.AbstractChannel$AbstractUnsafe.safeSetFailure(AbstractChannel.java:993) ~[netty-transport-4.1.50.Final.jar:4.1.50.Final]
        at io.netty.channel.AbstractChannel$AbstractUnsafe.bind(AbstractChannel.java:552) ~[netty-transport-4.1.50.Final.jar:4.1.50.Final]
        at io.netty.channel.DefaultChannelPipeline$HeadContext.bind(DefaultChannelPipeline.java:1334) ~[netty-transport-4.1.50.Final.jar:4.1.50.Final]
        at io.netty.channel.AbstractChannelHandlerContext.invokeBind(AbstractChannelHandlerContext.java:506) ~[netty-transport-4.1.50.Final.jar:4.1.50.Final]
        at io.netty.channel.AbstractChannelHandlerContext.bind(AbstractChannelHandlerContext.java:491) ~[netty-transport-4.1.50.Final.jar:4.1.50.Final]
        at io.netty.channel.DefaultChannelPipeline.bind(DefaultChannelPipeline.java:973) ~[netty-transport-4.1.50.Final.jar:4.1.50.Final]
        at io.netty.channel.AbstractChannel.bind(AbstractChannel.java:248) ~[netty-transport-4.1.50.Final.jar:4.1.50.Final]
        at io.netty.bootstrap.AbstractBootstrap$2.run(AbstractBootstrap.java:356) ~[netty-transport-4.1.50.Final.jar:4.1.50.Final]
        at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:164) ~[netty-common-4.1.50.Final.jar:4.1.50.Final]
        at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:472) ~[netty-common-4.1.50.Final.jar:4.1.50.Final]
        at io.netty.channel.epoll.EpollEventLoop.run(EpollEventLoop.java:387) ~[netty-transport-native-epoll-4.1.50.Final-linux-x86_64.jar:4.1.50.Final]
        at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989) ~[netty-common-4.1.50.Final.jar:4.1.50.Final]
        at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74) ~[netty-common-4.1.50.Final.jar:4.1.50.Final]
        at java.lang.Thread.run(Unknown Source) ~[?:?]
Caused by: io.netty.channel.unix.Errors$NativeIoException: bind(..) failed: Address already in use
```

Update 1:

I changed the ports to different numbers (because 8443 is used for both command and internal APIs - just different hostname) and I noticed that where it says 0.0.0.0:8443 already in use, the port it is trying to use is ZEEBE_BROKER_NETWORK_INTERNALAPI_ADVERTISEDPORT. I can understand that gateways listen to brokers on port 26502 just like brokers listen to other brokers on this port. The issue appears to be related to this. Still no solution to this error. Both gateways and brokers listen on port 26502,

I wonder how this is handled in code when one is using the embedded broker approach: how does the embedded gateway talk to the broker if both of them use the same port?

Update 2:

I am ruling out the gateway being the culprit. I have set different host and cluster.host values for the gateway (127.0.0.1) and the 0.0.0.0:8443 message persists. I think the bug is within the internalApi network configuration. In particular, I suspect the bug might be in the following function:

```
 public void applyDefaults(final NetworkCfg networkCfg) {
    advertisedHost =
        Optional.ofNullable(advertisedHost)
            .orElseGet(() -> Optional.ofNullable(host).orElseGet(networkCfg::getAdvertisedHost));
    host = Optional.ofNullable(host).orElse(networkCfg.getHost());
    port = Optional.ofNullable(port).orElse(defaultPort) + networkCfg.getPortOffset() * 10;
    advertisedPort =
        Optional.ofNullable(advertisedPort)
            .map(p -> p + networkCfg.getPortOffset() * 10)
            .orElse(port);
 }
```

I suspect that somehow the port and advertisedPort are taking the value of 8443, and host and advertisedHost are both taking 0.0.0.0 for reasons I don't understand, even though in the configuration I never set 0.0.0.0 for any hosts.

--- a/atomix/cluster/src/main/java/io/atomix/cluster/messaging/MessagingService.java
+++ b/atomix/cluster/src/main/java/io/atomix/cluster/messaging/MessagingService.java
@@ -18,6 +18,7 @@ package io.atomix.cluster.messaging;
 
 import io.atomix.utils.net.Address;
 import java.time.Duration;
+import java.util.Collection;
 import java.util.concurrent.CompletableFuture;
 import java.util.concurrent.Executor;
 import java.util.function.BiConsumer;
@@ -27,11 +28,19 @@ import java.util.function.BiFunction;
 public interface MessagingService {
 
   /**
-   * Returns the local messaging service address.
+   * Returns the local messaging service address. This is the address used by other nodes to
+   * communicate to this service.
    *
-   * @return the local address
+   * @return the address remote nodes use to communicate to this node
    */
   Address address();
+
+  /**
+   * Returns the interfaces to which the local messaging service is bind.
+   *
+   * @return the address the messaging service is bound to
+   */
+  Collection<Address> bindingAddresses();
 
   /**
    * Sends a message asynchronously to the specified communication address. The message is specified

--- a/atomix/cluster/src/main/java/io/atomix/cluster/messaging/impl/NettyMessagingService.java
+++ b/atomix/cluster/src/main/java/io/atomix/cluster/messaging/impl/NettyMessagingService.java
@@ -19,7 +19,6 @@ package io.atomix.cluster.messaging.impl;
 import static io.atomix.utils.concurrent.Threads.namedThreads;
 
 import com.google.common.base.Throwables;
-import com.google.common.collect.Lists;
 import com.google.common.collect.Maps;
 import com.google.common.util.concurrent.MoreExecutors;
 import io.atomix.cluster.messaging.ManagedMessagingService;
@@ -52,6 +51,8 @@ import io.netty.channel.socket.nio.NioSocketChannel;
 import io.netty.util.concurrent.Future;
 import java.net.InetAddress;
 import java.time.Duration;
+import java.util.ArrayList;
+import java.util.Collection;
 import java.util.Iterator;
 import java.util.List;
 import java.util.Map;
@@ -69,6 +70,7 @@ import java.util.concurrent.atomic.AtomicLong;
 import java.util.function.BiConsumer;
 import java.util.function.BiFunction;
 import java.util.function.Function;
+import java.util.stream.Collectors;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
@@ -76,9 +78,9 @@ import org.slf4j.LoggerFactory;
 public class NettyMessagingService implements ManagedMessagingService {
 
   private final Logger log = LoggerFactory.getLogger(getClass());
-  private final Address returnAddress;
+  private final Address advertisedAddress;
+  private final Collection<Address> bindingAddresses = new ArrayList<>();
   private final int preamble;
-  private final MessagingConfig config;
   private final ProtocolVersion protocolVersion;
   private final AtomicBoolean started = new AtomicBoolean(false);
   private final HandlerRegistry handlers = new HandlerRegistry();
@@ -95,33 +97,52 @@ public class NettyMessagingService implements ManagedMessagingService {
   private final List<CompletableFuture> openFutures;
 
   public NettyMessagingService(
-      final String cluster, final Address address, final MessagingConfig config) {
-    this(cluster, address, config, ProtocolVersion.latest());
+      final String cluster, final Address advertisedAddress, final MessagingConfig config) {
+    this(cluster, advertisedAddress, config, ProtocolVersion.latest());
   }
 
   NettyMessagingService(
       final String cluster,
-      final Address address,
+      final Address advertisedAddress,
       final MessagingConfig config,
       final ProtocolVersion protocolVersion) {
     preamble = cluster.hashCode();
-    returnAddress = address;
-    this.config = config;
+    this.advertisedAddress = advertisedAddress;
     this.protocolVersion = protocolVersion;
     openFutures = new CopyOnWriteArrayList<>();
     channelPool = new ChannelPool(this::openChannel, config.getConnectionPoolSize());
+    initAddresses(config);
+  }
+
+  private void initAddresses(final MessagingConfig config) {
+    final int port = config.getPort() != null ? config.getPort() : advertisedAddress.port();
+    if (config.getInterfaces().isEmpty()) {
+      bindingAddresses.add(Address.from(advertisedAddress.host(), port));
+    } else {
+      final List<Address> addresses =
+          config.getInterfaces().stream()
+              .map(iface -> Address.from(iface, port))
+              .collect(Collectors.toList());
+      bindingAddresses.addAll(addresses);
+    }
   }
 
   @Override
   public Address address() {
-    return returnAddress;
+    return advertisedAddress;
+  }
+
+  @Override
+  public Collection<Address> bindingAddresses() {
+    return bindingAddresses;
   }
 
   @Override
   public CompletableFuture<Void> sendAsync(
       final Address address, final String type, final byte[] payload, final boolean keepAlive) {
     final long messageId = messageIdGenerator.incrementAndGet();
-    final ProtocolRequest message = new ProtocolRequest(messageId, returnAddress, type, payload);
+    final ProtocolRequest message =
+        new ProtocolRequest(messageId, advertisedAddress, type, payload);
     return executeOnPooledConnection(
         address, type, c -> c.sendAsync(message), MoreExecutors.directExecutor());
   }
@@ -167,7 +188,8 @@ public class NettyMessagingService implements ManagedMessagingService {
     }
 
     final long messageId = messageIdGenerator.incrementAndGet();
-    final ProtocolRequest message = new ProtocolRequest(messageId, returnAddress, type, payload);
+    final ProtocolRequest message =
+        new ProtocolRequest(messageId, advertisedAddress, type, payload);
     if (keepAlive) {
       return executeOnPooledConnection(
           address, type, c -> c.sendAndReceive(message, timeout), executor);
@@ -237,7 +259,7 @@ public class NettyMessagingService implements ManagedMessagingService {
   @Override
   public CompletableFuture<MessagingService> start() {
     if (started.get()) {
-      log.warn("Already running at local address: {}", returnAddress);
+      log.warn("Already running at local address: {}", advertisedAddress);
       return CompletableFuture.completedFuture(this);
     }
 
@@ -370,7 +392,7 @@ public class NettyMessagingService implements ManagedMessagingService {
       final Function<ClientConnection, CompletableFuture<T>> callback,
       final Executor executor,
       final CompletableFuture<T> future) {
-    if (address.equals(returnAddress)) {
+    if (address.equals(advertisedAddress)) {
       callback
           .apply(localConnection)
           .whenComplete(
@@ -435,7 +457,7 @@ public class NettyMessagingService implements ManagedMessagingService {
       final Function<ClientConnection, CompletableFuture<T>> callback,
       final Executor executor) {
     final CompletableFuture<T> future = new CompletableFuture<>();
-    if (address.equals(returnAddress)) {
+    if (address.equals(advertisedAddress)) {
       callback
           .apply(localConnection)
           .whenComplete(
@@ -574,12 +596,9 @@ public class NettyMessagingService implements ManagedMessagingService {
    */
   private CompletableFuture<Void> bind(final ServerBootstrap bootstrap) {
     final CompletableFuture<Void> future = new CompletableFuture<>();
-    final int port = config.getPort() != null ? config.getPort() : returnAddress.port();
-    if (config.getInterfaces().isEmpty()) {
-      bind(bootstrap, Lists.newArrayList("0.0.0.0").iterator(), port, future);
-    } else {
-      bind(bootstrap, config.getInterfaces().iterator(), port, future);
-    }
+
+    bind(bootstrap, bindingAddresses.iterator(), future);
+
     return future;
   }
 
@@ -587,33 +606,28 @@ public class NettyMessagingService implements ManagedMessagingService {
    * Recursively binds the given bootstrap to the given interfaces.
    *
    * @param bootstrap the bootstrap to bind
-   * @param ifaces an iterator of interfaces to which to bind
-   * @param port the port to which to bind
+   * @param addressIterator an iterator of Addresses to which to bind
    * @param future the future to completed once the bootstrap has been bound to all provided
    *     interfaces
    */
   private void bind(
       final ServerBootstrap bootstrap,
-      final Iterator<String> ifaces,
-      final int port,
+      final Iterator<Address> addressIterator,
       final CompletableFuture<Void> future) {
-    if (ifaces.hasNext()) {
-      final String iface = ifaces.next();
+    if (addressIterator.hasNext()) {
+      final Address address = addressIterator.next();
       bootstrap
-          .bind(iface, port)
+          .bind(address.host(), address.port())
           .addListener(
               (ChannelFutureListener)
                   f -> {
                     if (f.isSuccess()) {
-                      log.info("TCP server listening for connections on {}:{}", iface, port);
+                      log.info("TCP server listening for connections on {}", address);
                       serverChannel = f.channel();
-                      bind(bootstrap, ifaces, port, future);
+                      bind(bootstrap, addressIterator, future);
                     } else {
                       log.warn(
-                          "Failed to bind TCP server to port {}:{} due to {}",
-                          iface,
-                          port,
-                          f.cause());
+                          "Failed to bind TCP server to port {} due to {}", address, f.cause());
                       future.completeExceptionally(f.cause());
                     }
                   });
@@ -695,7 +709,7 @@ public class NettyMessagingService implements ManagedMessagingService {
         final ChannelHandlerContext context,
         final Connection<M> connection,
         final ProtocolVersion protocolVersion) {
-      final MessagingProtocol protocol = protocolVersion.createProtocol(returnAddress);
+      final MessagingProtocol protocol = protocolVersion.createProtocol(advertisedAddress);
       context.pipeline().remove(this);
       context.pipeline().addLast("encoder", protocol.newEncoder());
       context.pipeline().addLast("decoder", protocol.newDecoder());

--- a/atomix/cluster/src/test/java/io/atomix/cluster/messaging/impl/NettyMessagingServiceTest.java
+++ b/atomix/cluster/src/test/java/io/atomix/cluster/messaging/impl/NettyMessagingServiceTest.java
@@ -29,11 +29,13 @@ import com.google.common.util.concurrent.MoreExecutors;
 import com.google.common.util.concurrent.Uninterruptibles;
 import io.atomix.cluster.messaging.ManagedMessagingService;
 import io.atomix.cluster.messaging.MessagingConfig;
+import io.atomix.cluster.messaging.MessagingService;
 import io.atomix.utils.net.Address;
 import io.zeebe.test.util.socket.SocketUtil;
 import java.net.ConnectException;
 import java.time.Duration;
 import java.util.Arrays;
+import java.util.List;
 import java.util.UUID;
 import java.util.concurrent.CompletableFuture;
 import java.util.concurrent.CompletionException;
@@ -69,6 +71,7 @@ public class NettyMessagingServiceTest {
   Address addressv21;
   Address addressv22;
   Address invalidAddress;
+  private MessagingService messagingService;
 
   @Before
   public void setUp() throws Exception {
@@ -423,5 +426,23 @@ public class NettyMessagingServiceTest {
     nettyv22.registerHandler(subject, (address, bytes) -> CompletableFuture.completedFuture(bytes));
     response = nettyv12.sendAndReceive(addressv22, subject, payload).get(10, TimeUnit.SECONDS);
     assertArrayEquals(payload, response);
+  }
+
+  @Test
+  public void shouldNotBindToAdvertisedAddress() {
+    // given
+    final var bindingAddress = Address.from(SocketUtil.getNextAddress().getPort());
+    final MessagingConfig config = new MessagingConfig();
+    config.setInterfaces(List.of(bindingAddress.host()));
+    config.setPort(bindingAddress.port());
+
+    // when
+    final Address nonBindableAddress = new Address("invalid.host", 1);
+    final var startFuture = new NettyMessagingService("test", nonBindableAddress, config).start();
+
+    // then - should not fail by using advertisedAddress for binding
+    messagingService = startFuture.join();
+    assertThat(messagingService.bindingAddresses()).contains(bindingAddress);
+    assertThat(messagingService.address()).isEqualTo(nonBindableAddress);
   }
 }

--- a/atomix/cluster/src/test/java/io/atomix/cluster/messaging/impl/TestMessagingService.java
+++ b/atomix/cluster/src/test/java/io/atomix/cluster/messaging/impl/TestMessagingService.java
@@ -27,6 +27,8 @@ import io.atomix.utils.concurrent.Futures;
 import io.atomix.utils.net.Address;
 import java.net.ConnectException;
 import java.time.Duration;
+import java.util.Collection;
+import java.util.List;
 import java.util.Map;
 import java.util.Set;
 import java.util.concurrent.CompletableFuture;
@@ -96,6 +98,11 @@ public class TestMessagingService implements ManagedMessagingService {
   @Override
   public Address address() {
     return address;
+  }
+
+  @Override
+  public Collection<Address> bindingAddresses() {
+    return List.of(address);
   }
 
   @Override

--- a/atomix/core/src/test/java/io/atomix/core/test/messaging/TestMessagingService.java
+++ b/atomix/core/src/test/java/io/atomix/core/test/messaging/TestMessagingService.java
@@ -27,6 +27,8 @@ import io.atomix.utils.concurrent.Futures;
 import io.atomix.utils.net.Address;
 import java.net.ConnectException;
 import java.time.Duration;
+import java.util.Collection;
+import java.util.List;
 import java.util.Map;
 import java.util.Set;
 import java.util.concurrent.CompletableFuture;
@@ -96,6 +98,11 @@ public class TestMessagingService implements ManagedMessagingService {
   @Override
   public Address address() {
     return address;
+  }
+
+  @Override
+  public Collection<Address> bindingAddresses() {
+    return List.of(address);
   }
 
   @Override

--- a/broker/src/main/java/io/zeebe/broker/Broker.java
+++ b/broker/src/main/java/io/zeebe/broker/Broker.java
@@ -8,6 +8,7 @@
 package io.zeebe.broker;
 
 import io.atomix.cluster.MemberId;
+import io.atomix.cluster.messaging.ManagedMessagingService;
 import io.atomix.cluster.messaging.MessagingConfig;
 import io.atomix.cluster.messaging.impl.NettyMessagingService;
 import io.atomix.core.Atomix;
@@ -32,6 +33,7 @@ import io.zeebe.broker.system.configuration.BrokerCfg;
 import io.zeebe.broker.system.configuration.ClusterCfg;
 import io.zeebe.broker.system.configuration.DataCfg;
 import io.zeebe.broker.system.configuration.NetworkCfg;
+import io.zeebe.broker.system.configuration.SocketBindingCfg;
 import io.zeebe.broker.system.configuration.backpressure.BackpressureCfg;
 import io.zeebe.broker.system.management.BrokerAdminService;
 import io.zeebe.broker.system.management.BrokerAdminServiceImpl;
@@ -206,7 +208,10 @@ public final class Broker implements AutoCloseable {
     startContext.addStep("actor scheduler", this::actorSchedulerStep);
     startContext.addStep("membership and replication protocol", () -> atomixCreateStep(brokerCfg));
     startContext.addStep(
-        "command api transport", () -> commandApiTransportStep(clusterCfg, localBroker));
+        "command api transport",
+        () ->
+            commandApiTransportStep(
+                clusterCfg, brokerCfg.getNetwork().getCommandApi(), localBroker));
     startContext.addStep(
         "command api handler", () -> commandApiHandlerStep(brokerCfg, localBroker));
     startContext.addStep("subscription api", () -> subscriptionAPIStep(localBroker));
@@ -273,25 +278,35 @@ public final class Broker implements AutoCloseable {
   }
 
   private AutoCloseable commandApiTransportStep(
-      final ClusterCfg clusterCfg, final BrokerInfo localBroker) {
-
-    final var nettyMessagingService =
-        new NettyMessagingService(
-            clusterCfg.getClusterName(),
-            Address.from(localBroker.getCommandApiAddress()),
-            new MessagingConfig());
-
-    nettyMessagingService.start().join();
-    LOG.debug("Bound command API to {} ", nettyMessagingService.address());
+      final ClusterCfg clusterCfg,
+      final SocketBindingCfg commpandApiConfig,
+      final BrokerInfo localBroker) {
+    final var messagingService = createMessagingService(clusterCfg, commpandApiConfig);
+    messagingService.start().join();
+    LOG.debug(
+        "Bound command API to {}, using advertised address {} ",
+        messagingService.bindingAddresses(),
+        messagingService.address());
 
     final var transportFactory = new TransportFactory(scheduler);
     serverTransport =
-        transportFactory.createServerTransport(localBroker.getNodeId(), nettyMessagingService);
+        transportFactory.createServerTransport(localBroker.getNodeId(), messagingService);
 
     return () -> {
       serverTransport.close();
-      nettyMessagingService.stop().join();
+      messagingService.stop().join();
     };
+  }
+
+  private ManagedMessagingService createMessagingService(
+      final ClusterCfg clusterCfg, final SocketBindingCfg socketCfg) {
+    final var messagingConfig = new MessagingConfig();
+    messagingConfig.setInterfaces(List.of(socketCfg.getHost()));
+    messagingConfig.setPort(socketCfg.getPort());
+    return new NettyMessagingService(
+        clusterCfg.getClusterName(),
+        Address.from(socketCfg.getAdvertisedHost(), socketCfg.getAdvertisedPort()),
+        messagingConfig);
   }
 
   private AutoCloseable commandApiHandlerStep(

--- a/broker/src/main/java/io/zeebe/broker/clustering/atomix/AtomixFactory.java
+++ b/broker/src/main/java/io/zeebe/broker/clustering/atomix/AtomixFactory.java
@@ -71,6 +71,8 @@ public final class AtomixFactory {
             .withClusterId(clusterCfg.getClusterName())
             .withMemberId(localMemberId)
             .withMembershipProtocol(membershipProtocol)
+            .withMessagingInterface(networkCfg.getInternalApi().getHost())
+            .withMessagingPort(networkCfg.getInternalApi().getPort())
             .withAddress(
                 Address.from(
                     networkCfg.getInternalApi().getAdvertisedHost(),

--- a/parent/pom.xml
+++ b/parent/pom.xml
@@ -574,6 +574,12 @@
       </dependency>
 
       <dependency>
+        <groupId>org.testcontainers</groupId>
+        <artifactId>toxiproxy</artifactId>
+        <version>${version.testcontainers}</version>
+      </dependency>
+
+      <dependency>
         <groupId>com.github.docker-java</groupId>
         <artifactId>docker-java-api</artifactId>
         <version>${version.docker-java-api}</version>

--- a/qa/integration-tests/pom.xml
+++ b/qa/integration-tests/pom.xml
@@ -130,6 +130,11 @@
     </dependency>
 
     <dependency>
+      <groupId>org.testcontainers</groupId>
+      <artifactId>toxiproxy</artifactId>
+    </dependency>
+
+    <dependency>
       <groupId>io.zeebe</groupId>
       <artifactId>zeebe-protocol-asserts</artifactId>
       <scope>test</scope>

--- a/qa/integration-tests/src/test/java/io/zeebe/broker/it/network/AdvertisedAddressTest.java
+++ b/qa/integration-tests/src/test/java/io/zeebe/broker/it/network/AdvertisedAddressTest.java
@@ -1,0 +1,163 @@
+/*
+ * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH under
+ * one or more contributor license agreements. See the NOTICE file distributed
+ * with this work for additional information regarding copyright ownership.
+ * Licensed under the Zeebe Community License 1.0. You may not use this file
+ * except in compliance with the Zeebe Community License 1.0.
+ */
+package io.zeebe.broker.it.network;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import io.zeebe.client.ZeebeClient;
+import io.zeebe.client.ZeebeClientBuilder;
+import io.zeebe.client.api.response.Topology;
+import io.zeebe.containers.ZeebeBrokerContainer;
+import io.zeebe.containers.ZeebeGatewayContainer;
+import io.zeebe.containers.ZeebePort;
+import io.zeebe.containers.ZeebeTopologyWaitStrategy;
+import io.zeebe.test.util.asserts.TopologyAssert;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.slf4j.LoggerFactory;
+import org.testcontainers.containers.GenericContainer;
+import org.testcontainers.containers.Network;
+import org.testcontainers.containers.ToxiproxyContainer;
+import org.testcontainers.utility.DockerImageName;
+
+/**
+ * Use ToxiProxy. Even though it does not support UDP, we can still use ToxiProxy because only
+ * "gossip" messages use UDP. SWIM has other messages to probe and sync that uses TCP. So the
+ * brokers can still find each other.
+ */
+public class AdvertisedAddressTest {
+  private static final String TOXIPROXY_NETWORK_ALIAS = "toxiproxy";
+  private static final String TOXIPROXY_IMAGE = "shopify/toxiproxy:2.1.0";
+  private static final int CLUSTER_SIZE = 3;
+  private static final int PARTITION_COUNT = 1;
+  private static final int REPLICATION_FACTOR = 3;
+  private static final String ZEEBE_IMAGE_VERSION = "camunda/zeebe:current-test";
+  @Rule public final Network network = Network.newNetwork();
+
+  @Rule
+  public final ToxiproxyContainer toxiproxy =
+      new ToxiproxyContainer(DockerImageName.parse(TOXIPROXY_IMAGE))
+          .withNetwork(network)
+          .withNetworkAliases(TOXIPROXY_NETWORK_ALIAS);
+
+  private List<ZeebeBrokerContainer> containers;
+  private List<String> initialContactPoints;
+  private ZeebeGatewayContainer gateway;
+
+  @Before
+  public void setup() {
+    initialContactPoints = new ArrayList<>();
+    containers =
+        IntStream.range(0, CLUSTER_SIZE)
+            .mapToObj(i -> new ZeebeBrokerContainer(ZEEBE_IMAGE_VERSION))
+            .collect(Collectors.toList());
+
+    gateway = new ZeebeGatewayContainer(ZEEBE_IMAGE_VERSION);
+    IntStream.range(0, CLUSTER_SIZE).forEach(i -> configureBrokerContainer(i, containers));
+    configureGatewayContainer(gateway, initialContactPoints.get(0));
+  }
+
+  @After
+  public void tearDown() {
+    containers.parallelStream().forEach(GenericContainer::stop);
+  }
+
+  @Test
+  public void shouldCommunicateOverProxy() {
+    // given
+    containers.parallelStream().forEach(GenericContainer::start);
+    gateway.start();
+
+    // when
+    final ZeebeClientBuilder zeebeClientBuilder =
+        ZeebeClient.newClientBuilder()
+            .usePlaintext()
+            .gatewayAddress(gateway.getExternalGatewayAddress());
+    final Topology topology;
+    try (final var client = zeebeClientBuilder.build()) {
+      topology = client.newTopologyRequest().send().join(5, TimeUnit.SECONDS);
+      // then - can find each other
+      TopologyAssert.assertThat(topology).isComplete(3, 1);
+
+      // when
+      final var messageSend =
+          client
+              .newPublishMessageCommand()
+              .messageName("test")
+              .correlationKey("test")
+              .send()
+              .join(5, TimeUnit.SECONDS);
+      // then - gateway can talk to the broker
+      assertThat(messageSend.getMessageKey()).isPositive();
+    }
+  }
+
+  private void configureBrokerContainer(final int index, final List<ZeebeBrokerContainer> brokers) {
+    final int clusterSize = brokers.size();
+    final var broker = brokers.get(index);
+    final var hostName = "broker-" + index;
+    final var commandApiProxy = toxiproxy.getProxy(hostName, ZeebePort.COMMAND.getPort());
+    final var internalApiProxy = toxiproxy.getProxy(hostName, ZeebePort.INTERNAL.getPort());
+    final var monitoringApiProxy = toxiproxy.getProxy(hostName, ZeebePort.MONITORING.getPort());
+
+    initialContactPoints.add(
+        TOXIPROXY_NETWORK_ALIAS + ":" + internalApiProxy.getOriginalProxyPort());
+
+    LoggerFactory.getLogger("Test")
+        .info("Configuring broker with initial contactpoints {}", initialContactPoints);
+
+    broker
+        .withNetwork(network)
+        .withNetworkAliases(hostName)
+        .withEnv("ZEEBE_BROKER_NETWORK_MAXMESSAGESIZE", "128KB")
+        .withEnv("ZEEBE_BROKER_CLUSTER_NODEID", String.valueOf(index))
+        .withEnv("ZEEBE_BROKER_CLUSTER_CLUSTERSIZE", String.valueOf(clusterSize))
+        .withEnv("ZEEBE_BROKER_CLUSTER_REPLICATIONFACTOR", String.valueOf(REPLICATION_FACTOR))
+        .withEnv("ZEEBE_BROKER_CLUSTER_PARTITIONCOUNT", String.valueOf(PARTITION_COUNT))
+        .withEnv(
+            "ZEEBE_BROKER_CLUSTER_INITIALCONTACTPOINTS", String.join(",", initialContactPoints))
+        .withEnv("ZEEBE_LOG_LEVEL", "DEBUG")
+        .withEnv("ATOMIX_LOG_LEVEL", "INFO")
+        .withEnv("ZEEBE_BROKER_NETWORK_COMMANDAPI_ADVERTISEDHOST", TOXIPROXY_NETWORK_ALIAS)
+        .withEnv(
+            "ZEEBE_BROKER_NETWORK_COMMANDAPI_ADVERTISEDPORT",
+            String.valueOf(commandApiProxy.getOriginalProxyPort()))
+        .withEnv("ZEEBE_BROKER_NETWORK_INTERNALAPI_ADVERTISEDHOST", TOXIPROXY_NETWORK_ALIAS)
+        .withEnv(
+            "ZEEBE_BROKER_NETWORK_INTERNALAPI_ADVERTISEDPORT",
+            String.valueOf(internalApiProxy.getOriginalProxyPort()))
+        .withEnv("ZEEBE_BROKER_NETWORK_MONITORINGAPI_ADVERTISEDHOST", TOXIPROXY_NETWORK_ALIAS)
+        .withEnv(
+            "ZEEBE_BROKER_NETWORK_MONITORINGAPI_ADVERTISEDPORT",
+            String.valueOf(monitoringApiProxy.getOriginalProxyPort()))
+        // Since gossip does not work with ToxiProxy, increase the sync interval so changes are
+        // propagated faster
+        .withEnv("ZEEBE_BROKER_CLUSTER_MEMBERSHIP_SYNCINTERVAL", "100ms");
+  }
+
+  private void configureGatewayContainer(
+      final ZeebeGatewayContainer gateway, final String initialContactPoint) {
+    // gateway is not behind proxy
+    gateway
+        .withEnv("ZEEBE_GATEWAY_CLUSTER_CONTACTPOINT", initialContactPoint)
+        .withTopologyCheck(
+            new ZeebeTopologyWaitStrategy()
+                .forBrokersCount(CLUSTER_SIZE)
+                .forPartitionsCount(PARTITION_COUNT)
+                .forReplicationFactor(REPLICATION_FACTOR))
+        .withNetwork(network)
+        .withNetworkAliases("gateway");
+  }
+}

